river/cluster/evostream.py

```diff
@@ -0,0 +1,424 @@
+
+        evolution_index = self.time_stamp % self.evolution_interval
+
+        if evolution_index == 0:
+            self._evolution_batch[self.evolution_interval - 1] = x
+        elif evolution_index == 1:
+            self._evolution_batch = {0: x}
+        else:
+            self._evolution_batch[evolution_index - 1] = x
+
+        temp_mc = evoStreamMicroCluster(x=x, last_update=self.time_stamp, weight=1)
+
+        merged_status = False
+
+        for micro_cluster in self._micro_clusters.values():
+            if self._distance(temp_mc.center, micro_cluster.center) < self.radius:
+                magnitude = self._gaussian_neighborhood(
+                    temp_mc.center, micro_cluster.center
+                )
+                for i in micro_cluster.center:
+                    micro_cluster.center[i] += magnitude * (
+                        temp_mc.center[i] - micro_cluster.center[i]
+                    )
+                micro_cluster.last_update = self.time_stamp
+                micro_cluster.weight = (
+                    micro_cluster.weight
+                    * (
+                        2
+                        ** (
+                            -self.decay_rate
+                            * (self.time_stamp - micro_cluster.last_update)
+                        )
+                    )
+                    + 1
+                )
+                merged_status = True
+
+        if not merged_status:
+            self._micro_clusters[len(self._micro_clusters)] = temp_mc
+
+    def _fitness(self, points, clusters):
+        ssq = 0
+        for i, point_i in points.items():
+            closest_cluster_index_i = self._find_closest_cluster_index(
+                point_i, clusters
+            )
+            distance = self._distance(point_i, clusters[closest_cluster_index_i].center)
+            ssq += distance * distance
+        return 1 / ssq
+
+    def _roulette_wheel_selection(self, fitness_clusters):
+        sum_fitness = sum(fitness_clusters.values())
+        pick_1, pick_2 = random.uniform(0, sum_fitness), random.uniform(0, sum_fitness)
+        key_1, key_2 = -1, -1
+        current = 0
+
+        for key, fitness in fitness_clusters.items():
+            current += fitness
+            if current > pick_1 and key_1 == -1:
+                key_1 = key
+            if current > pick_2 and key_2 == -1:
+                key_2 = key
+            if key_1 != -1 and key_2 != -1:
+                break
+
+        return self._cluster_population[key_1], self._cluster_population[key_2]
+
+    def _evolution(self):
+        self._fitness_cluster_population = {
+            i: self._fitness(self._evolution_batch, self._cluster_population[i])
+            for i in self._cluster_population
+        }
+
+        p_1, p_2 = self._roulette_wheel_selection(self._fitness_cluster_population)
+        o_1, o_2 = self._binary_crossover(p_1, p_2)
+        delta = random.uniform(0, 1)
+
+        def mutate(g):
```

Comment:
What about using a `@staticmethod` instead?

Reply:
@staticmethod used. I thought it would have been better to leave the function within the original `_evolution` function, but now thinking back, pulling it outside as a static method would leave the code look nicer!